httpcore5/src/main/java/org/apache/hc/core5/http/protocol/ForwardedRequest.java

```diff
@@ -0,0 +1,189 @@
+ * The {@link EndpointDetails} is used to determine the by and for key values. If a Forwarded header
+ * already exists in the request, the existing header is not overwritten; instead, the new header
+ * value is appended to the existing header value, with a comma separator.
+ * <p>
+ * This implementation of the ForwardedRequest class is immutable and thread-safe.
+ *
+ * @since 5.3
+ */
+@Contract(threading = ThreadingBehavior.IMMUTABLE)
+public class ForwardedRequest implements HttpRequestInterceptor {
+
+    /**
+     * The name of the header to set in the HTTP request.
+     */
+    private static final String FORWARDED_HEADER_NAME = "Forwarded";
+
+    /**
+     * Singleton instance.
+     */
+    public static final HttpRequestInterceptor INSTANCE = new ForwardedRequest();
+
+
+    /**
+     * The process method adds a Forwarded header to an HTTP request containing information about
+     * the intermediate nodes that the request has passed through. If a Forwarded header already
+     * exists in the request, the new header value is appended to the existing header value, with a
+     * comma separator.
+     * <p>
+     * The method retrieves the {@link ProtocolVersion} and {@link URIAuthority} from the
+     * HttpContext. The ProtocolVersion is used to determine the proto key value and the
+     * URIAuthority is used to determine the host and port key values. The method also retrieves the
+     * EndpointDetails from the {@link HttpContext}, if it exists. The {@link EndpointDetails} is
+     * used to determine the by key value.
+     *
+     * @param request the HTTP request to which the Forwarded header is to be added (not
+     *                {@code null})
+     * @param entity  the entity details of the request (can be {@code null})
+     * @param context the HTTP context in which the request is being executed (not {@code null})
+     * @throws HttpException     if there is an error processing the request
+     * @throws IOException       if an IO error occurs while processing the request
+     * @throws ProtocolException if the request authority is not specified
+     */
+    @Override
+    public void process(final HttpRequest request, final EntityDetails entity, final HttpContext context) throws HttpException, IOException {
+        Args.notNull(request, "HTTP request");
+        Args.notNull(context, "HTTP context");
+
+        final ProtocolVersion ver = context.getProtocolVersion() != null ? context.getProtocolVersion() : HttpVersion.HTTP_1_1;
+
+        final URIAuthority authority = request.getAuthority();
+        if (authority == null) {
+            throw new ProtocolException("Request authority not specified");
+        }
+
+        final int port = authority.getPort();
+
+        final StringBuilder valueBuilder = new StringBuilder();
+        final String forwardedHeaderValue = request.getFirstHeader(FORWARDED_HEADER_NAME) != null ?
```

Comment:
@arturobernalg Maybe the compiler can optimize it, but I doubt it. So help it. Use a variable to store the result of `request.getFirstHeader(FORWARDED_HEADER_NAME)`